core: introduce NUMAPolicy and NUMAMask options

msekletar opened the pull request:
Make possible to set NUMA allocation policy for manager. Manager's
policy is by default inherited to all forked off processes. However, it
is possible to override the policy on per-service basis. Currently we
support, these policies: default, prefer, bind, interleave,local.
See man set_mempolicy for details on each policy.

Overall NUMA policy actually consists of two parts. Policy itself and
bitmask representing NUMA nodes where is policy effective. Node mask can
be specified using related option, NUMAMask. Again, default mask can be
overwritten on per-service level.

msekletar commented:
I also wrote an integration test for the features which can be later merged to CentOS CI (it uses beakerlib). 

https://github.com/msekletar/systemd-NUMAPolicy-test/blob/master/runtest.sh

And here is an output from the test run (caveat: I've used patched strace so it properly recognizes MPOL_LOCAL argument for set_mempolicy),

https://gist.github.com/msekletar/366942c8978d7a870c5f789226a7adec

msekletar commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,103 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+                goto out;
+
+        bits = policy->nodes.allocated * 8;
+        ulong_bits = sizeof(unsigned long) * 8;
```
Not sure I grok this.

keszybz replied:
You're calling `new0` with `bits / ulong_bits` rounded up. I think it'd be easier to use `DIV_ROUND_UP(bits, ulong_bits)` instead of `(bits / ulong_bits) + (bits % ulong_bits  > 0)`.

msekletar commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,103 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+
+out:
+        *nodes = TAKE_PTR(out);
+        *maxnode = num_numa_nodes() + 1;
```
I mean in theory it would be fine to use the code as is because it doesn't make sense to set bits in a mask higher than maximum count of NUMA nodes on the box.

keszybz replied:
Let's say I'm on a system with 8192 CPUs, but `policy->nodes.allocated` is 128. Then `*maxnode` would become 8193, but `*nodes*` would point to a buffer of 128 bytes or . The man page says that "nodemask points to a bit mask of node IDs that contains up to maxnode bits". The syscall would then fail with EFAULT.

> Also it appears that first thing that kernel does when it tries to copy bitmask from userspace is that it decrements maxnode.
>
> https://github.com/torvalds/linux/blob/2f4c53349961c8ca480193e47da4d44fdb8335a8/mm/mempolicy.c#L1286

Duh, so it seems. That seems broken or at least contradictory to the man page.

msekletar commented:
@keszybz I've updated the PR. PTAL. 

Btw +1 for maxnode argument of set_mempolicy is something that I actually discovered in libnuma code when I was wondering why is kernel returning -EINAL to me.

https://github.com/numactl/numactl/blob/1959c774398fa6859ff98dde76c01c7579be2731/libnuma.c#L293

msekletar commented:
Also it appears that first thing that kernel does when it tries to copy bitmask from userspace is that it decrements maxnode.

https://github.com/torvalds/linux/blob/2f4c53349961c8ca480193e47da4d44fdb8335a8/mm/mempolicy.c#L1286

I was running into problems on 4 NUMA node system where I was trying to set BIND policy for the 4th node (node 3). Kernel was giving me an error until I bumped the maxnode to 5.

keszybz commented on NEWS:
```diff
@@ -104,6 +104,12 @@ CHANGES WITH 243 in spe:
+          hence will set the default policy for PID1. Default policy can be
+          overriden on per-service basis. Related setting NUMAMask= is used to
+          specify NUMA node mask that should be associated with the selected
+          policy
```
Missing ".".

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,103 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+        }
+
+        return n_nodes;
+}
```
So this is unused now, no?

msekletar commented:
@keszybz I've fixed the issues and force pushed my branch. PTAL 

semaphoreci failed but I have no idea why.

keszybz commented on src/core/dbus-execute.c:
```diff
@@ -225,6 +225,45 @@ static int property_get_cpu_affinity(
+        assert(reply);
+        assert(c);
+
+        return sd_bus_message_append_basic(reply, 'i', &c->numa_policy.type);
```
I think it'd be better to pass the policy over dbus as a string ("default", "bind", ...). It would be much more user-friendly and would make introspection and use from scripts easier to understand.

msekletar replied:
I discussed this with @poettering and his guidance was to pass everything over dbus in parsed form, i.e. no stringy APIs. I don't care either way, but you guys should agree on general guidance.

keszybz replied:
@poettering?

poettering replied:
hmm, difficult question. Generally, when we have enums with number assignments defined by others that are API, then we should just use the numbers (example: waitid() CLD_XXX constants, or unix SIGXXX signal constants). If we have internal stuff however, or any other stuff where the numbers really shouldn't be exposed, then we shouldn't expose them, and use strings instead. 

And then there are things in between: sometimes we want to use official stuff, but with special "invalid"  or "default" values. in that case strings might be better, as then we could use the empty string for that.

So dunno, it really depends on what this enum is like here. 

If in doubt we should probably opt for strings rather than integers...

keszybz commented on src/core/main.c:
```diff
@@ -1753,6 +1803,21 @@ static void update_cpu_affinity(bool skip_setup) {
 
+static void update_numa_policy(bool skip_setup) {
+        _cleanup_free_ char *nodes = NULL;
+        char const * policy = NULL;
```
`const char *policy`.

keszybz commented on src/core/main.c:
```diff
@@ -1753,6 +1803,21 @@ static void update_cpu_affinity(bool skip_setup) {
+        if (skip_setup || !numa_policy_type_valid(arg_numa_policy.type))
+                return;
+
+        policy = numa_policy_type_to_string(arg_numa_policy.type);
```
Please put this under `if (DEBUG_LOGGING) {`.

mrc0mmand commented:
After like eleventh attempt I finally managed to get the logs (thanks, JavaScript) and the build just time outed, so it doesn't seem to be relevant to this PR. In any case, I triggered a rebuild, hopefully a successful one this time.

mrc0mmand commented:
https://salsa.debian.org/ci-team/autopkgtest.git returns 503, so Semaphore is now out of the game (cc @mbiebl)

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,85 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+        assert(nodes);
+
+        /* Take a shortcut here if policy is set to default or local since kernel requires that we pass NULL as nodemask */
+        if (policy->type == NUMA_POLICY_DEFAULT || policy->type == NUMA_POLICY_LOCAL)
```
IN_SET()...

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,85 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+
+        assert(policy);
+
+        if (get_mempolicy(NULL, NULL, 0, 0, 0) < 0 && errno == ENOSYS)
```
Hmm, why is this check needed?

msekletar replied:
This is not strictly needed. Idea here is to have an "early return" on machines which don't support NUMA.

I mean conversion between parsed configuration and actual representation expected by ```set_mempolicy()``` calls ```malloc()```. I think it is better if we avoid it knowing that NUMA is not supported.

keszybz replied:
I'm was wondering whether it's worth the extra syscall. But OK, seems reasonable to leave this in.

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,85 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+
+        r = set_mempolicy(mode, nodes, maxnode);
+        if (r < 0)
+                return r;
```
```C
if (set_mempolicy(...) < 0)
    return -errno;
```

msekletar replied:
Duh...

keszybz commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,85 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+        if (!out)
+                return -ENOMEM;
+
+        for (node = 0; node < bits; node++)
```
I think it would be nice to add comment here why this conversion is needed, since quite often the input and output will be exactly the same.

msekletar commented:
@keszybz updated. PTAL.

keszybz commented:
LGTM, apart from the int vs. string interface thingy. Let's wait for @poettering's reply on that.

mrc0mmand commented on src/core/main.c:
```diff
@@ -582,6 +583,53 @@ static int config_parse_cpu_affinity2(
         return 0;
 }
 
+static int config_parse_numa_policy(
```
A somewhat tangential note: would it make sense to move the `core/main.c`-specific `config_parse` functions (not only the NUMA ones, but the rest of them as well) to, for example, `shared/config-parser.*`, so a fuzzer could be written around the config-parsing machinery (mainly the separate `config_parse*` functions)? @keszybz

keszybz replied:
Yes, it would. Do you want to do this?

mrc0mmand replied:
Sure, will look into that.

poettering commented on src/basic/missing_syscall.h:
```diff
@@ -444,3 +444,46 @@ static inline ssize_t missing_statx(int dfd, const char *filename, unsigned flag
+#  ifdef __NR_get_mempolicy
+        i = syscall(__NR_get_mempolicy, mode, nodemask, maxnode, addr, flags);
+#  else
+       errno = ENOSYS;
```
something is off with the indentation in this function, i.e. the `errno = …`is aligned differently than the `i = …` two lines earlier

poettering commented on src/core/execute.c:
```diff
@@ -3148,6 +3148,14 @@ static int exec_child(
+                r = apply_numa_policy(&context->numa_policy);
+                if (r < 0) {
+                        *exit_status = EXIT_NUMA_MEM_POLICY;
+                        return log_unit_error_errno(unit, errno, "Failed to set NUMA memory policy: %m");
```
the error code is in "r" here, not in "errno"

poettering commented on src/core/execute.c:
```diff
@@ -4336,6 +4344,14 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
+
+                nodes = cpu_set_to_range_string(&c->numa_policy.nodes);
+                fprintf(f, "%sNUMAPolicy: %s\n", prefix, numa_policy_type_to_string(c->numa_policy.type));
+                fprintf(f, "%sNUMAMask: %s\n", prefix, nodes);
```
use strnull(nodes) or so, in case the cpu_set_to_range_string() memory allocation failed). in some cases printf/fprintf doesn't like NULL at all for %s, hence let's avoid it

poettering commented on src/core/load-fragment.c:
```diff
@@ -1211,6 +1211,56 @@ int config_parse_exec_cpu_sched_policy(const char *unit,
+        }
+
+        c->numa_policy.type = policy;
+
```
please use `DEFINE_CONFIG_PARSE_ENUM` to define this one

poettering commented on src/core/main.c:
```diff
@@ -582,6 +583,53 @@ static int config_parse_cpu_affinity2(
+        (void) parse_cpu_set_extend(rvalue, &policy->nodes, true, unit, filename, line, lvalue);
+
+        return 0;
+}
```
these two functions we already basically have, in load-fragment.c. please reuse.

poettering commented on src/core/main.c:
```diff
@@ -1753,6 +1803,23 @@ static void update_cpu_affinity(bool skip_setup) {
+        if (DEBUG_LOGGING) {
+                policy = numa_policy_type_to_string(arg_numa_policy.type);
+                nodes = cpu_set_to_range_string(&arg_numa_policy.nodes);
+                log_debug("Setting NUMA policy to %s, with nodes %s.", strnull(policy), !isempty(nodes) ? nodes : "(null)");
```
use strnull() for the ternary op

poettering commented on src/core/main.c:
```diff
@@ -1753,6 +1803,23 @@ static void update_cpu_affinity(bool skip_setup) {
+        }
+
+        if (apply_numa_policy(&arg_numa_policy) < 0)
+                log_warning_errno(errno, "Failed to set NUMA memory policy: %m");
```
this returns the error in its return value, no? not in errno

poettering commented on src/shared/bus-unit-util.c:
```diff
@@ -1049,6 +1049,58 @@ static int bus_append_execute_property(sd_bus_message *m, const char *field, con
+
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
```
you can do all of the above in one call:

```c
r = sd_bus_message_append(m, "(sv)", field, "i", (int32_t) r);
```

poettering replied:
this has not been addressed yet

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -7,12 +7,20 @@
+#include "dirent-util.h"
+#include "errno-util.h"
+#include "fd-util.h"
 #include "extract-word.h"
```
plese keep includes ordered alphabetically

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,87 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+
+        /* Take a shortcut here if policy is set to default or local since kernel requires that we pass NULL as nodemask */
+        if (IN_SET(policy->type, NUMA_POLICY_DEFAULT, NUMA_POLICY_LOCAL))
+                goto out;
```
if this branch is taken then `*maxnode` will be 1 afterwards, even though *nodes is NULL. Does that make sense?

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,87 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
         return 0;
 }
+
+static int numa_policy_to_mempolicy(const NUMAPolicy *policy, unsigned long *maxnode, unsigned long **nodes) {
```
i think it would make sense to prefix the latter two params with `ret_` to clarify that they are output params

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,87 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+                break;
+        default:
+                assert_not_reached("Invalid NUMA policy");
+        }
```
so this mapping is the identity, no? why do we need a separate enum for this then, if we just map basically the same names on themselves?

poettering commented on src/shared/cpu-set-util.h:
```diff
@@ -48,3 +48,36 @@ int cpu_set_to_dbus(const CPUSet *set, uint8_t **ret, size_t *allocated);
+        NUMA_POLICY_LOCAL,
+        _NUMA_POLICY_MAX,
+        _NUMA_POLICY_INVALID = -1,
+} NUMAPolicyType;
```
this type appears redundant, no? sure, it adds a type name to the set of MPOL_xyz values defined in missing_syscall.h, but other than that it doesn't do much, no? i.e. we don't add anything useful here (besides the typedef name), hence I think we should probably not have this, but just use the kernels own type here (i.e. `int`), even if it's not as typesafe.

msekletar replied:
Well, we use the type in bunch of places. We have translation table for to/from string conversions defined on it, accompanying helper function prototypes and such.

I think it is cleaner from the design perspective to have two types separated and have just a single place where we translate from our representation to low level kernel type, even when map operation is trivial.

poettering replied:
but why have both sets of enums? it's not worth it. Yes, it would have been great if the kernel's own definition would have used a proper enum type for this, but they don't. Let's accept that, and simple use the type they picked. In no other case we have second guessed the kernel's type and introduced our own type that the same but also not. I mean, it's stupid that on UNIX fds are "int"s, instead of their own type fd_t. But it's how it is, we have to follow it, we don't define our own fd_t, but just use "int", because that's what the kernel returns.

Note that the string translation stuff works the same for enums and for true ints. We use that various places, look at sched_policy_to_string() and friends for example. In fact, the sched policy stuff is almost the same as the numa policy case, it's also an int, with only a handful of values defined. Please copy the same concepts, including adding mpol_is_valid() or so, copying what sched_policy_is_valid() currently provides...

poettering commented on src/shared/cpu-set-util.h:
```diff
@@ -48,3 +48,36 @@ int cpu_set_to_dbus(const CPUSet *set, uint8_t **ret, size_t *allocated);
+}
+
+int apply_numa_policy(const NUMAPolicy *policy);
+unsigned num_numa_nodes(void);
```
this prototype is not actually used, no?

poettering commented on src/shared/exit-status.c:
```diff
@@ -157,6 +157,9 @@ const char* exit_status_to_string(int status, ExitStatusLevel level) {
                         return "CONFIGURATION_DIRECTORY";
 
+                case EXIT_NUMA_MEM_POLICY:
+                        return "NUMA";
```
please keep the enum name and the string returned in sync

poettering replied:
not addressed yet

msekletar commented on man/systemd.exec.xml:
```diff
@@ -761,6 +761,28 @@ CapabilityBoundingSet=~CAP_B CAP_C</programlisting>
+        <listitem><para>Controls the NUMA node list which will be applied alongside with selected NUMA policy.
+        Takes a list of NUMA nodes and has the same syntax as a list of CPUs for <varname>CPUAffinity=</varname>
+        option. Note that the list of NUMA nodes is not required for <option>default</option> and <option>local</option>
+        policies and for <option>prefer</option> policy we expect a single NUMA node.</para></listitem>
```
I don't think it would be a good idea to pick policy automatically. Imagine I have a NUMAMask with just a single node and no specification for NUMAPolicy. I wouldn't know which policy to pick as admin might wanted strict policy (MPOL_BIND) however more relaxed policy (MPOL_PREFERRED) makes sense for mask with just one node too.

poettering replied:
Well, but the user can always pick the policy explicitly if they want control. I think there's nothing wrong with deriving some roughly sensible automatic defaults from settings if things are underspecified.

I mean, let's turn this around: unless I am mistaken using MPOL_DEFAULT but specifying a mask makes no sense whatsoever. Hence, if the user only specifies a mask, why not default to MPOL_PREFERRED or so (which appears to be a good default to me if a mask is specified)? we should make this all work with minimal effort if we can. i.e. make the common stuff easy and the complex stuff possible...

msekletar replied:
If the policy is configured improperly (underspecified) I'd rather fail service startup instead of picking something and letting it run potentially misconfigured (in a sense that applied policy would violate user's expectations, even though he made a mistake and left things misconfigured he arguably had some expectations about what the outcome should be like).

I mean this stuff is going to be used only by handful of services. And will be always explicit user config. Distros will never ship unit files with NUMA stuff enabled by default. So I don't think that we need to be too user-friendly here. Moreover, I'd bet you that this user friendly defaults would be considered a bug by the crowd that is target audience for this feature. 

If you really insist and won't press the green button w/o this then I am going to do it but I don't agree I must say.

@keszybz any opinions ?

keszybz replied:
I agree with @poettering here. `MPOL_PREFERRED` seems to be the correct choice almost always, so if user specifies the mask, we should default to that.

Werkov commented:
Hi (I hope I didn't miss any answers before).
1) What is the intended behavior on systems that don't support `set_mempolicy`? E.g., systemd built without `HAVE_SET_MEMPOLICY`. It seems to me that even units that don't specify any particular NUMA policy will fail to start with `EXIT_NUMA_MEM_POLICY`. (My vote is not to do any mempolicy stuff for exec contexts that don't specify it explicitly.)

Nitpicks

2) `struct NUMAPolicy` embeds `CPUSet`, strictly speaking NUMA nodes != cpus, wouldn't be a generic name like bitmask better?

3) <del>What about hard coding (config option) maximum supported number of CPUs/NUMA nodes and hence avoid dynamic allocations?</del> It's in every ExecContext, that might be too big memory footprint if it should follow kernels NR_CPUS.

msekletar commented:
> 1. What is the intended behavior on systems that don't support `set_mempolicy`? E.g., systemd built without `HAVE_SET_MEMPOLICY`. It seems to me that even units that don't specify any particular NUMA policy will fail to start with `EXIT_NUMA_MEM_POLICY`. (My vote is not to do any mempolicy stuff for exec contexts that don't specify it explicitly.)

It will be handled this way on the unit level. Policy will be reset, by default, to special "invalid" value indicating that it wasn't set and we won't attempt to apply any policy if policy has that value (-1 in my latest version which I haven't pushed yet).
 
> * `struct NUMAPolicy` embeds `CPUSet`, strictly speaking NUMA nodes != cpus, wouldn't be a generic name like bitmask better?

`CPUSet` contains `cpu_set_t` hence the name. `cpu_set_t` type from glibc could have been named differently. We can change the name later on, if it proves to be a problem. 

> 2\. It's in every ExecContext, that might be too big memory footprint if it should follow kernels NR_CPUS.

But we allocate NUMA node mask only if it is specified in the unit file. Also number of NUMA nodes is usually much much smaller than number of CPUs.

poettering commented:
> It will be handled this way on the unit level. Policy will be reset, by default, to special "invalid" value indicating that it wasn't set and we won't attempt to apply any policy if policy has that value (-1 in my latest version which I haven't pushed yet).
> 
> > * ``

i think we should always keep the numa bits in place but when we try to apply them an the kernel doesn't have support for this (i.e. ENOSYS is returned) we should just debug log about that and continue. i.e. apply when it makes sense, but when it really doesn't then gracefully continue but make it debuggable via debug logging.

poettering commented:
>     2\. It's in every ExecContext, that might be too big memory footprint if it should follow kernels NR_CPUS.

It's just one ptr, one int and one size_t per unit, it's not precisely the world I think. Note that the bitmask itself is not embedded into the Unit, only a ptr to it. And if it's not set then nothing beyond the ptr itself is allocated.

msekletar commented:
Updated my branch.

In the latest revision I got rid of `NUMAPolicyType` enum as requested by @poettering and also implemented a new default behaviour which sets allocation policy to `MPOL_PREFERRED` if `NUMAMask=` was specified (albeit it must contain only single NUMA node because this is an expectation from the kernel) and `NUMAPolicy=` wasn't.

@poettering @keszybz PTAL. 

Thank you for your reviews and feedback so far!

poettering commented on src/core/execute.c:
```diff
@@ -3148,6 +3148,16 @@ static int exec_child(
+        if (mpol_is_valid(context->numa_policy.type)) {
+                r = apply_numa_policy(&context->numa_policy);
+                if (r == -EOPNOTSUPP)
+                        log_unit_debug_errno(unit, SYNTHETIC_ERRNO(EOPNOTSUPP), "NUMA support not available, ignoring.");
```
the error is not synthethic here, you just received it. Just replace `SYNTHETIC_ERRNO(EOPNOTSUPP)` with a simple `r`

poettering commented on src/core/load-fragment.c:
```diff
@@ -1211,6 +1212,35 @@ int config_parse_exec_cpu_sched_policy(const char *unit,
+
+        r = parse_cpu_set_extend(rvalue, &p->nodes, true, unit, filename, line, lvalue);
+        if (r < 0) {
+                log_syntax(unit, LOG_ERR, filename, line, 0, "Failed to parse NUMA node mask, ignoring: %s", rvalue);
```
why pass the error here as `0`? pass `r` here, there's no reason not to pass it to log_syntax(). (Note that log_syntax will include the error code in a structured logging field ERRNO= if it has it, even if `%m` is not used in the message format string)

poettering commented on src/core/main.c:
```diff
@@ -1753,6 +1756,27 @@ static void update_cpu_affinity(bool skip_setup) {
+
+        r = apply_numa_policy(&arg_numa_policy);
+        if (r == -EOPNOTSUPP)
+                log_debug("NUMA support not available, ignoring.");
```
if you have an error code, always pass it to the logging function, regardless if you include it in the msg text or not. hence:

`log_debug()` → `log_debug_errno(r,` in this case

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,98 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+          This function should be called after NUMAMask= was parsed. If the NUMAPolicy= is parsed
+          later it will override the default that we set here. */
+        if (policy->nodes.allocated != 0 && policy->type == -1)
+                policy->type = MPOL_PREFERRED;
```
please don't actually update the field here. This means the information that we automatically picked the preferred setting is lost that way, and this matters for example if inside the same unit file you first set a mask, and later unset it again. the first setting will fix the policy to preferred, and that will stick even though setting and then unsetting a unit option really should undo everything.

My recommendation: never update the NUMAPolicy structure with the automatically determined value. Instead, define an accessor function:

 ```c
int numa_policy_get_type(NUMAPolicy *policy) {
        assert(policy);
        return policy->type < 0 ? (policy->nodes.set ? MPOL_PREFERRED : MPOL_DEFAULT) : policy->type;
}
```

And then, when ever you want to know the numa policy type just write `numa_policy_get_type(p)` instead of `p->type`

msekletar replied:
We can't do that I think. Looking at `execute.c` where we apply the policy for new units,

```C
        if (mpol_is_valid(context->numa_policy.type)) {
                r = apply_numa_policy(&context->numa_policy);
                if (r == -EOPNOTSUPP)
                        log_unit_debug_errno(unit, SYNTHETIC_ERRNO(r), "NUMA support not available, ignoring.");
                else if (r < 0) {
                        *exit_status = EXIT_NUMA_POLICY;
                        return log_unit_error_errno(unit, r, "Failed to set NUMA memory policy: %m");
                }
        }
```

If I take your suggestion then we would violate the current semantics. Because we don't want to reset policy to default if none was set. We want to inherit policy from the manager.

poettering replied:
then maybe change the accessor to:

```c
return policy->type < 0 ? (policy->nodes.set ? MPOL_PREFERRED : -1) : policy->type;
```

i.e. propagate the -1 if no nodes are set, so that you then know it's not set.

msekletar replied:
Yes, that should work

poettering commented on src/shared/cpu-set-util.c:
```diff
@@ -287,3 +295,98 @@ int cpu_set_from_dbus(const uint8_t *bits, size_t size, CPUSet *set) {
+
+static const char* const mpol_table[] = {
+        [MPOL_DEFAULT]    = "default",
+        [MPOL_PREFERRED]  = "prefer",
```
i think we should stick to the kernel's own names here. If the enum value is called MPOL_PREFERRED we should use the string `"preferred"` too, not `"prefer"`

msekletar commented:
Updated. @poettering PTAL.

keszybz commented on src/core/main.c:
```diff
@@ -1753,6 +1756,27 @@ static void update_cpu_affinity(bool skip_setup) {
+static void update_numa_policy(bool skip_setup) {
+        int r;
+        _cleanup_free_ char *nodes = NULL;
+        const char * policy = NULL;
```
Please move those defined under the DEBUG_LOGGING conditional.

poettering replied:
these two are no used outside of the `if (DEBUG_LOGGING)`, please move them into that scope hence. (usually we try to define variables at minimal scope only, except for `int r`)

poettering commented on src/core/dbus-execute.c:
```diff
@@ -1687,6 +1732,20 @@ int bus_exec_context_set_transient_property(
+                if (r < 0)
+                        return r;
+
+                if (!mpol_is_valid(type))
```
thinking about this, we should accept type < 0 here too, so that this can be reset again

msekletar replied:
Hmm, but ExecContext properties can't be set later after a transient unit was started.

poettering replied:
well, but we allow in unit files to say this:

```
…
[Service]
NUMAMask=1 2 3 4
NUMAMask=5 6 7 8
NUMAMask=
```

hence i think we should try to also make this work:

systemd-run -p NUMMask="1 2 3 4" -p NUMAMask="5678" -p NUMAMask="" …

And yes, this has value. For example, nspawn also takes the same --property= params, and passes them to machined, machined then adds a few more settings (but prefixes them, so that the nspawn supplied ones take precedence) and passes it on to PID 1 for scope registration. In that case it might be useful to cancel out the props machined set with the ones specified on the nspawn cmdline, and we generally can do that, and hence should also support that for the NUMMask and NUMAPolicy...

poettering commented on src/core/execute.c:
```diff
@@ -3148,6 +3148,16 @@ static int exec_child(
+        if (mpol_is_valid(numa_policy_get_type(&context->numa_policy))) {
+                r = apply_numa_policy(&context->numa_policy);
+                if (r == -EOPNOTSUPP)
+                        log_unit_debug_errno(unit, SYNTHETIC_ERRNO(r), "NUMA support not available, ignoring.");
```
the error is not synthetic here (we are not generating it here, but received it already from the previous call). Please drop the SYNTHETIC_ERRNO() wrapper here, just use `r` directly.

msekletar replied:
Hmm, but `apply_numa_policy()` does produce value synthetically if we detect that syscall is not supported.

keszybz replied:
SYNTHETIC_ERRNO() just means "generated here", nothing more. Anything which is passed down from a called function should be just logged normally.

msekletar replied:
Ahh, ok. I will send followup then.
